Fix #25350 - Some JS removes State/Province field for United Kingdom for example

Bartlomiejsz opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This is fix for #25350 

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#25350: Some JS removes State/Province field for United Kingdom for example

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Click Address Book
2. Select the United Kingdom as a country
3. Set Postcode to AB12 3CD
4. Set State/Province to Greater London
5. Set City to London
6. Save address.
7. Verify that you see "Greater London" in Default Shipping and Default Billing address
8. Click Change Billing Address

#### Expected result (*)
Greater London is in the State/Province input

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

krzksz reviewed (changes requested):
Hey @Bartlomiejsz, thanks for the contribution! I have trouble to understand the extensiveness of this change, I can see you are also refactoring helpers but was it really required to touch PHP files at all? I did some short debugging and following changes inside `_updateRegion` method solved the issue for me and cleaned up the code a bit:

1.  Move `regionInput.val('');` below `if (this.options.regionJson[country]) {` as this clearing input field every time.
2. Extract `this._removeSelectOptions(regionList);` above `if...else` because it is present in both of them anyway.
3. Remove `$(regionList).find('option:selected').removeAttr('selected');` as all the options are removed by the above call anyway.

Can you please check if those changes would be ok for you once you apply them? If so, feel free to update the PR.

Bartlomiejsz commented:
Hi @krzksz, I'll check those changes as soon as possible, now only regarding php files being changed - this indeed wasn't required in terms of functional changes here and in first version of PR those were not modified. I modified those later to fix failing static tests ($this shouldn't be used in phtml files)

Bartlomiejsz commented:
Hi @krzksz, I applied changes you requested, please review :)

krzksz commented on app/code/Magento/Checkout/view/frontend/web/js/region-updater.js:
```diff
@@ -42,6 +48,10 @@ define([
+                this.currentRegionValue[this.element.val()] = $(e.target).val();
+            }, this));
+
             $(this.options.regionInputId).on('focusout', $.proxy(function () {
```
Only one remark from my side: `$(this.options.regionInputId)` is now called twice, could you either introduce `regionInput` variable at the top of the function and reuse it or chain the `.on` calls together? I don't lean towards any of the solutions, decide which looks cleaner yourself.

Bartlomiejsz commented:
Hi @krzksz, you're right, added this one change. I also noticed one additional issue, if you select some country which have regions list as select (i.e. USA), select one of regions, then switch to another country with also regions select (like Poland). Then Region select becomes empty (even without default message to choose option). I modified then saving of current region in same manner as introduced in this PR for text input - per country.

magento-engcom-team commented:
Hi @krzksz, thank you for the review.
[ENGCOM-6318](https://jira.corp.magento.com/browse/ENGCOM-6318) has been created to process this Pull Request
:eight_spoked_asterisk: @krzksz, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

engcom-Bravo commented:
Hello @krzksz. Would You please mark this PR with an appropriate [label](https://github.com/magento/magento2/pull/25386#issuecomment-557722395). Thank You.

krzksz commented:
@engcom-Bravo Done.

engcom-Bravo commented:
Hello @Bartlomiejsz. Thank You for Your contribution. In order for us to be able to further process Your PR please cover it with the appropriate auto-tests. Thank You.

Bartlomiejsz commented:
Hi @engcom-Bravo, sorry but what exactly should be covered with tests here? In my previous PRs all javascript changes were marked as `Auto Tests: Not Required`, ViewModel I added is simply returning existing helpers and it's addition I already included in `dev/tests/integration/testsuite/Magento/Customer/Block/Form/RegisterTest.php`.

krzksz commented:
It would be nice to have a test that covers the scenario you gave and makes sure that values are properly preserved.

Bartlomiejsz commented:
Hi @krzksz, if you mean scenario in js script then unfortunately I don't have experience with writing tests for javascript, so not able to do so at the moment, and also I believe tests weren't requirement for js so far:
https://github.com/magento/magento2/pull/24833
https://github.com/magento/magento2/pull/24255

krzksz commented:
Having code changes covered by tests is a part of Magento DoD: https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html#automated-tests

I'm not sure if it is possible to properly test your changes using frontend unit tests, but maybe adding a functional one would be an option. I also don't have any experience with MFTF, but a few of my MRs are blocked by the lack of them so there always have to be first time, right? 😄

Bartlomiejsz commented:
@krzksz you aren't wrong for sure :smile: 
Probably js test would fit here better than unit/functional, so I'll try to work on that

engcom-Echo commented:
@magento run all tests

Bartlomiejsz committed:
Fix #25350 - Some JS removes State/Province field for United Kingdom for example

Bartlomiejsz commented:
Hi @krzksz, since https://github.com/magento/magento2/pull/26285 introduced ViewModel used for similar goal that I achieved with one created here, and both changes together resulted in conflicts, I rebased my changes onto 2.4-develop and instead of creating my own ViewModel I used one created in mentioned PR. Please review again if possible

engcom-Echo commented:
I will take care of test coverage.

ptylek commented:
@Bartlomiejsz this branch has conflicts that must be resolved

Bartlomiejsz commented:
@ptylek done, resolved
